### PR TITLE
LP-2824 Populate course status fields in the ApplicationHub model

### PR DESCRIPTION
**Pull request by javeriaejaz-arbisoft**

### CODE REVIEW

**Review by aightmunam — CHANGES_REQUESTED** [review]

File: openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        not completed the business line and common business line courses
+
+        Returns:
+            A querySet of user ids whose business line and common business line pre req courses
+            are not yet marked as completed
+        """
+        return User.objects.filter(
+            Q(application_hub__is_written_application_completed=True) &
+            Q(application_hub__program_prereq_courses_status=ApplicationHub.COMPLETED) &
+            ~Q(application_hub__bu_prereq_courses_status=ApplicationHub.COMPLETED)
```
[the remark this answers is not in the document]
Comment (javeriaejaz-arbisoft) [reply openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py:57]: As per the flow, business line courses are available to user after application submission and program pre-reqs courses. So, to filter out the maximum users, we have placed both checks. Therefore, when user complete bu courses as well, a congratulation email is sent to him/her.

File: openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        Returns:
+            users_with_pre_reqs_completed(list): filtered users that have completed for program pre-reqs
+        """
+        users_with_pre_reqs_completed = []
+
+        for user in users_with_pre_reqs_in_progress:
+            courses = MultilingualCourseGroup.objects.get_user_program_prereq_courses(user)
+            if has_user_passed_given_courses(user, courses):
+                users_with_pre_reqs_completed.append(user)
```
[the remark this answers is not in the document]
Comment (javeriaejaz-arbisoft) [reply openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py:70]: True but the above is more readable than this:
`users_with_pre_reqs_completed.append(user) if has_user_passed_given_courses(user, courses) else None`
And below isn't recommended style.
`if has_user_passed_given_courses(user, courses): users_with_pre_reqs_completed.append(user)`

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2535/workflows/workflows_id/jobs/2535/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/populate_course_status_data.py (68.0%): Missing lines 30-32,51-52,63-64,67
openedx/adg/lms/applications/management/commands/tests/test_populate_course_status_data.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_bu_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_program_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py (90.0%): Missing lines 41-42,45
openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py (92.3%): Missing lines 40-41
openedx/adg/lms/applications/managers.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/applications/tests/test_views.py (100%)
openedx/adg/lms/courseware_override/tests/test_helpers.py (100%)
-------------
Total:   224 lines
Missing: 13 lines
Coverage: 94%
-------------
```

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2538/workflows/workflows_id/jobs/2538/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/populate_course_status_data.py (66.7%): Missing lines 25-27,46-47,58-59,62
openedx/adg/lms/applications/management/commands/tests/test_populate_course_status_data.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_bu_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_program_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py (90.0%): Missing lines 40-41,44
openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py (92.3%): Missing lines 39-40
openedx/adg/lms/applications/managers.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/applications/tests/test_views.py (100%)
openedx/adg/lms/courseware_override/tests/test_helpers.py (100%)
-------------
Total:   218 lines
Missing: 13 lines
Coverage: 94%
-------------
```

File: openedx/adg/lms/applications/management/commands/populate_course_status_data.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        Set the value of prerequisite courses status to IN_PROGRESS for users who are
+        actively enrolled in any pre_req course.
+
+        Arguments:
+            user_ids_with_application_hub (list): list of user ids
+            is_bu_prereq (bool): boolean flag to update status for either program pre_reqs or business line pre_reqs
+        """
+        field_to_update = 'bu_prereq_courses_status' if is_bu_prereq else 'program_prereq_courses_status'
+        users_with_prereqs_enrolled = get_users_with_pre_reqs_enrolled(
```
Comment (harris-irfan) [inline openedx/adg/lms/applications/management/commands/populate_course_status_data.py:59]: `users_enrolled_in_prereqs` would be more appropriate.

File: openedx/adg/lms/applications/management/commands/populate_course_status_data.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            user_ids_with_application_hub, is_bu_prereq
+        )
+        bulk_update_application_hub_field(
+            field_to_update, ApplicationHub.IN_PROGRESS, users_with_prereqs_enrolled
+        )
+
+    def set_prereqs_courses_status_to_completed(self, is_bu_prereq=False):
```
Comment (harris-irfan) [inline openedx/adg/lms/applications/management/commands/populate_course_status_data.py:66]: `set_prereq_courses_status_to_completed`

**Review by harris-irfan — CHANGES_REQUESTED** [review]

**Comment by harris-irfan** [comment]
Unit tests are yet to be reviewed.

**Review by harris-irfan — COMMENTED** [review]
Unit tests review

File: openedx/adg/lms/applications/management/commands/populate_course_status_data.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        Update the program_prereq_courses_status/bu_prereq_courses_status field in Application Hub
+
+        Arguments:
+            is_bu_prereq (bool): if True updates 'bu_prereq_courses_status' otherwise 'program_prereq_courses_status'
+        """
+        if is_bu_prereq:
+            ApplicationHub.objects.filter(is_bu_prerequisite_courses_passed=True).update(
```
[the remark this answers is not in the document]
Comment (javeriaejaz-arbisoft) [reply openedx/adg/lms/applications/management/commands/populate_course_status_data.py:74]: But filter is using two different flags too.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2548/workflows/workflows_id/jobs/2548/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/populate_course_status_data.py (66.7%): Missing lines 25-27,46-47,58-59,62
openedx/adg/lms/applications/management/commands/tests/test_populate_course_status_data.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_bu_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_program_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py (90.0%): Missing lines 40-41,44
openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py (92.3%): Missing lines 39,42
openedx/adg/lms/applications/managers.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/applications/tests/test_views.py (100%)
openedx/adg/lms/courseware_override/tests/test_helpers.py (100%)
-------------
Total:   223 lines
Missing: 13 lines
Coverage: 94%
-------------
```

**Review by harris-irfan — APPROVED** [review]
Awesome work done on this ticket!! 👏

File: openedx/adg/lms/applications/helpers.py [hunk 6]
```diff
@@ -738,3 +741,46 @@ def create_html_string_for_course_scores_in_admin_review(user_application):
...
+        users_enrolled_in_bu_prereqs (list): list of users
+    """
+    from openedx.adg.lms.applications.models import MultilingualCourseGroup
+
+    prereq_course_groups = (
+        MultilingualCourseGroup.objects.program_prereq_course_groups() if not update_bu_prereq
+        else MultilingualCourseGroup.objects.business_line_and_common_business_line_prereq_course_groups()
+    )
+
+    users_enrolled_in_bu_prereqs = get_users_with_active_enrollments_from_course_groups(
```
Comment (aightmunam) [inline openedx/adg/lms/applications/helpers.py:783]: Rename this to `users_enrolled_in_prereqs` as we are considering both bu and program courses

**Review by aightmunam — APPROVED** [review]
LGTM! 👏🏼

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2551/workflows/workflows_id/jobs/2551/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/populate_course_status_data.py (66.7%): Missing lines 25-27,46-47,58-59,62
openedx/adg/lms/applications/management/commands/tests/test_populate_course_status_data.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_bu_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_program_prereq_courses_status.py (100%)
openedx/adg/lms/applications/management/commands/update_bu_prereq_courses_status.py (90.0%): Missing lines 40-41,44
openedx/adg/lms/applications/management/commands/update_program_prereq_courses_status.py (92.3%): Missing lines 39,42
openedx/adg/lms/applications/managers.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/applications/tests/test_views.py (100%)
openedx/adg/lms/courseware_override/tests/test_helpers.py (100%)
-------------
Total:   223 lines
Missing: 13 lines
Coverage: 94%
-------------
```

**Pull request merged**